lightning/src/util/sweep.rs

```diff
@@ -0,0 +1,688 @@
+						&key,
+						false,
+					) {
+						Ok(_) => {
+							log_debug!(self.logger, "Pruning swept output as sufficiently confirmed via spend in transaction {:?}",
+								o.latest_spending_tx.as_ref().map(|t| t.txid()));
+							return false;
+						}
+						Err(e) => {
+							log_error!(
+								self.logger,
+								"Removal of key {}/{}/{} failed due to: {}",
+								SPENDABLE_OUTPUT_INFO_PERSISTENCE_PRIMARY_NAMESPACE,
+								SPENDABLE_OUTPUT_INFO_PERSISTENCE_SECONDARY_NAMESPACE,
+								key,
+								e
+							);
+							return true;
+						}
+					}
+				}
+			}
+			true
+		});
+	}
+
+	fn persist_info(&self, output: &TrackedSpendableOutput) -> Result<(), io::Error> {
+		let key = id_to_hex_string(&output.id);
+		let data = output.encode();
+		self.kv_store
+			.write(
+				SPENDABLE_OUTPUT_INFO_PERSISTENCE_PRIMARY_NAMESPACE,
+				SPENDABLE_OUTPUT_INFO_PERSISTENCE_SECONDARY_NAMESPACE,
+				&key,
+				&data,
+			)
+			.map_err(|e| {
+				log_error!(
+					self.logger,
+					"Write for key {}/{}/{} failed due to: {}",
+					SPENDABLE_OUTPUT_INFO_PERSISTENCE_PRIMARY_NAMESPACE,
+					SPENDABLE_OUTPUT_INFO_PERSISTENCE_SECONDARY_NAMESPACE,
+					key,
+					e
+				);
+				e
+			})
+	}
+}
+
+impl<B: Deref, ES: Deref, F: Deref, K: Deref, L: Deref> Listen for OutputSweeper<B, ES, F, K, L>
+where
+	B::Target: BroadcasterInterface,
+	ES::Target: EntropySource,
+	F::Target: Filter + Sync + Send,
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	fn filtered_block_connected(
+		&self, header: &Header, txdata: &chain::transaction::TransactionData, height: u32,
+	) {
+		{
+			let best_block = self.best_block.lock().unwrap();
+			assert_eq!(best_block.block_hash(), header.prev_blockhash,
+			"Blocks must be connected in chain-order - the connected header must build on the last connected header");
+			assert_eq!(best_block.height(), height - 1,
+			"Blocks must be connected in chain-order - the connected block height must be one greater than the previous height");
+		}
+
+		self.transactions_confirmed(header, txdata, height);
+		self.best_block_updated(header, height);
+	}
+
+	fn block_disconnected(&self, header: &Header, height: u32) {
+		let new_height = height - 1;
+		{
+			let mut best_block = self.best_block.lock().unwrap();
+			assert_eq!(best_block.block_hash(), header.block_hash(),
+				"Blocks must be disconnected in chain-order - the disconnected header must be the last connected header");
+			assert_eq!(best_block.height(), height,
+				"Blocks must be disconnected in chain-order - the disconnected block must have the correct height");
+			*best_block = BestBlock::new(header.prev_blockhash, new_height)
+		}
+
+		let mut locked_outputs = self.outputs.lock().unwrap();
+		for output_info in locked_outputs.iter_mut() {
+			if output_info.confirmation_hash == Some(header.block_hash()) {
+				debug_assert_eq!(output_info.confirmation_height, Some(height));
+				output_info.confirmation_hash = None;
+				output_info.confirmation_height = None;
```

Comment:
> Not quite sure I'm following along, isn't that what `SpendableOutputStatus` in the linked branch does? Note there is a N:M relationship from spendable output to broadcast transaction and a N:1 relationship from spendable output to confirmed transaction.

Ah, yeah the naming and hash maps threw me off. Should have took a closer look. Guess I don't know what you meant by "overhead/redundancy". Seems you can just take what you have in this PR and make an enum covering six of the `Option` fields.

> I mean, sure, we could (probably) rip it apart and puzzle it together in the serialization logic, but, although I originally favored the STM design, I'm not convinced that redoing it (again) is much less error prone than proceeding with the current version that by now had several eyes on it?
> 

I mean, from a reviewer, reader, and maintenance perspective, having three well-defined enum variants is much simpler to reason about than six `Option` fields. With the enum approach, the state transition is obvious and setting the correct fields is enforced by the compiler. With the `Option` approach, I need to piece together which state it is in and then confirm that all `Option` fields are set or cleared appropriately upon transitioning.

Maybe there's something about the enum approach that I'm not seeing that makes it less ideal?

> In any case, if you much prefer it I'll see if I can create a draft some time soonish. If we really want go that way, would you be okay with doing the refactor as a follow-up, as this PR already has a good bit of review on it and seems close to being merged otherwise?

Sure, I'm fine with a follow-up given this has already been reviewed for LDK Node.